pkg/kv/cosmosdb/store.go

```diff
@@ -0,0 +1,324 @@
+}
+
+func (s *Store) Get(ctx context.Context, partitionKey, key []byte) (*kv.ValueWithPredicate, error) {
+	if len(partitionKey) == 0 {
+		return nil, kv.ErrMissingPartitionKey
+	}
+	if len(key) == 0 {
+		return nil, kv.ErrMissingKey
+	}
+	item := Document{
+		PartitionKey: s.encoding.EncodeToString(partitionKey),
+		ID:           s.encoding.EncodeToString(key),
+	}
+	pk := azcosmos.NewPartitionKeyString(item.PartitionKey)
+
+	// Read an item
+	itemResponse, err := s.containerClient.ReadItem(ctx, pk, item.ID, nil)
+	if err != nil {
+		var respErr *azcore.ResponseError
+		if errors.As(err, &respErr) && respErr.StatusCode == http.StatusNotFound {
+			return nil, kv.ErrNotFound
+		}
+		return nil, err
+	}
+
+	var itemResponseBody Document
+	err = json.Unmarshal(itemResponse.Value, &itemResponseBody)
+	if err != nil {
+		return nil, err
+	}
+
+	val, err := s.encoding.DecodeString(itemResponseBody.Value)
+	if err != nil {
+		return nil, err
+	}
+	return &kv.ValueWithPredicate{
+		Value:     val,
+		Predicate: kv.Predicate([]byte(itemResponse.ETag)),
+	}, nil
+}
+
+func (s *Store) Set(ctx context.Context, partitionKey, key, value []byte) error {
+	if len(partitionKey) == 0 {
+		return kv.ErrMissingPartitionKey
+	}
+	if len(key) == 0 {
+		return kv.ErrMissingKey
+	}
+	if value == nil {
+		return kv.ErrMissingValue
+	}
+
+	// Specifies the value of the partiton key
+	item := Document{
+		PartitionKey: s.encoding.EncodeToString(partitionKey),
+		ID:           s.encoding.EncodeToString(key),
+		Value:        s.encoding.EncodeToString(value),
+	}
+
+	b, err := json.Marshal(item)
+	if err != nil {
+		return err
+	}
+	itemOptions := azcosmos.ItemOptions{
+		ConsistencyLevel: azcosmos.ConsistencyLevelBoundedStaleness.ToPtr(),
+	}
+	pk := azcosmos.NewPartitionKeyString(item.PartitionKey)
+
+	_, err = s.containerClient.UpsertItem(ctx, pk, b, &itemOptions)
+	return err
+}
+
+func (s *Store) SetIf(ctx context.Context, partitionKey, key, value []byte, valuePredicate kv.Predicate) error {
+	if len(partitionKey) == 0 {
+		return kv.ErrMissingPartitionKey
+	}
+	if len(key) == 0 {
+		return kv.ErrMissingKey
+	}
+	if value == nil {
+		return kv.ErrMissingValue
+	}
+
+	// Specifies the value of the partiton key
+	item := Document{
+		PartitionKey: s.encoding.EncodeToString(partitionKey),
+		ID:           s.encoding.EncodeToString(key),
+		Value:        s.encoding.EncodeToString(value),
+	}
+
+	b, err := json.Marshal(item)
+	if err != nil {
+		return err
+	}
+	itemOptions := azcosmos.ItemOptions{
+		ConsistencyLevel: azcosmos.ConsistencyLevelBoundedStaleness.ToPtr(),
+	}
+	pk := azcosmos.NewPartitionKeyString(item.PartitionKey)
+
+	if valuePredicate == nil {
+		_, err = s.containerClient.CreateItem(ctx, pk, b, &itemOptions)
+		var respErr *azcore.ResponseError
+		if errors.As(err, &respErr) && respErr.StatusCode == http.StatusConflict {
+			return kv.ErrPredicateFailed
+		}
+		return err
+	} else if valuePredicate == kv.PrecondConditionalExists {
```

Comment:
Nice!

Reply:
Sometimes Azure exposing 6 different APIs for doing the same thing is helpful.